### ISSUE TITLE
some questions/feedback about language server

### ISSUE THREAD

**Issue by sunyi000** [galaxyproject/galaxy-language-server#105]
I have tried a little bit of the VS code plugin.. The plugin is great, it saves lots of effort creating the xml wrapper

I'm still exploring it and I have the following questions/feedback so far.. 

1. Would it be possible to have a Sublime plugin as well? VS code is great but I believe there's a large population of people using sublime as their main editor.

2. When import macros from another file, in the <macros><import> element, there's a validation error says 
`Validation error on macro: Element 'tool': Character content other than whitespace is not allowed because the content type is 'element-only'.`  I don't quite understand the error. 'planemo lint' returns no issues.

3. If I start typing <param>from scratch, I see a list of valid attributes shows up, which is great. It is working as expected.

However, it seems to behave differently if I open an existing tool wrapper xml. The below 2 questions happened when opening an existing tool wrapper.

If I open an existing tool wrapper, and trying to see if the tool is valid, after putting my mouse cursor after <param and press space I was expecting to see the same list of valid attributes for <param, but instead, I see a different list "id, version, hidden, display_interface" etc... I thought this list is actually the attributes for tool, and thus caused some confusion.

4. if I press backspace after doing 3 above, I see a list of suggested galaxy param type, gx-boolean, gx-text etc. For gx-text, gx-integer, gx-float, 'argument' attribute appear first. 
I would thought for these three, 'argument' may be place at the last. It is optional and only used if the parameter reflects to a command line argument. Also, 'name' attribute is not in the suggested list for these three parameter types.

5. For the auto-generation of tests. It generate a basic test template, and all I have to do is to change the param values. However, for one of the tool I have, it generates the same test template 16 times by press ctl+alt+t .. not sure why 16, so some documentation on the auto-generation of test would be great. For example, if the tool has 5 parameters, the factorial of 5 could be more than 100, how do you control the number of test template to create?

Thanks very much for the great effort creating this language server. I will come back again if I have more questions.

**Comment by davelopez** [comment]
Many thanks for your feedback @sunyi000!!

>Would it be possible to have a Sublime plugin as well? VS code is great but I believe there's a large population of people using sublime as their main editor.

It should be possible since there seems to be a plugin to support the [Language Server Protocol](https://microsoft.github.io/language-server-protocol/) in Sublime [here](https://github.com/sublimelsp/LSP).
However, I haven't used Sublime myself so I may not be the right person to implement the client. Anyway, in theory, it would be similar to [one of these configurations](https://lsp.readthedocs.io/en/latest/#python) but using the [`galaxy-language-server`](https://pypi.org/project/galaxy-language-server/). But again, I am not familiar with Sublime so I may be wrong.

> When import macros from another file, in the element, there's a validation error says
Validation error on macro: Element 'tool': Character content other than whitespace is not allowed because the content type is 'element-only'. I don't quite understand the error. 'planemo lint' returns no issues.

There are some differences in how planemo and the GLS validates the XML that contains macros. The main reason behind this is that when using macros the document you see in the editor no longer matches (in terms of lines and offsets) with the real document that will be validated after expanding the macros. I agree the validation process using macros needs some improvements as right now is a simple check against the XSD schema so, in the meantime, always rely on planemo for the final verdict :)

> If I open an existing tool wrapper, and trying to see if the tool is valid, after putting my mouse cursor after <param and press space I was expecting to see the same list of valid attributes for <param, but instead, I see a different list "id, version, hidden, display_interface" etc... I thought this list is actually the attributes for tool, and thus caused some confusion.

This may very well be a bug. Can you tell me the exact version of the plugin you are using and provide a link to the actual tool wrapper so I can reproduce it and try to fix it? That would be great :)

> if I press backspace after doing 3 above, I see a list of suggested galaxy param type, gx-boolean, gx-text etc. For gx-text, gx-integer, gx-float, 'argument' attribute appear first.
I would thought for these three, 'argument' may be place at the last. It is optional and only used if the parameter reflects to a command line argument. Also, 'name' attribute is not in the suggested list for these three parameter types.

Those `gx-*something*` are code snippets, they are not to be used inside the tag definition (where the attributes are placed) they are meant to be used to quickly write new `<param>` and other common tags like this:

![issue](https://user-images.githubusercontent.com/46503462/106481410-b3e1bc00-64ac-11eb-9790-52c29005eb6a.gif)

I agree it may be confusing if they appear when your cursor context is inside the tag definition name. I will investigate to see if I can avoid showing them in those cases.

>For the auto-generation of tests. It generate a basic test template, and all I have to do is to change the param values. However, for one of the tool I have, it generates the same test template 16 times by press ctl+alt+t .. not sure why 16, so some documentation on the auto-generation of test would be great. For example, if the tool has 5 parameters, the factorial of 5 could be more than 100, how do you control the number of test template to create?

You are right, there is little to no documentation about how the tests or the command section is generated apart from the code itself [and some of the tests](https://github.com/galaxyproject/galaxy-language-server/tree/master/server/galaxyls/tests/files). I will try to add some more information in the readme. 
Answering your question, it basically generates a tree taking into account the `conditional` tags and their possible option values, and then creates a test case for each of the tree leaves. Of course, this could lead to potentially dozens of test cases depending on the complexity of your tool. You are free to remove and edit them as you desire since they are only meant to provide some scaffolding to save some typing and to check that 'roughly' every possible execution path that your tool may take has a test.

**Comment by sunyi000** [comment]
thanks for the detailed answers @davelopez 👍 

I used this tool I just created in the past days to explore the language server
https://github.com/sunyi000/galaxytools/blob/cellprofiler-extra4-mod/tools/cellprofiler/color_to_gray.xml

The plugin version is 0.3.2 as it is the default in the vs code markets. May be I did not use VScode properly, but the mains thing I found is that the plugin behaves differently when - type everything from scratch and
- open existing wrapper and try to edit..

For the auto-generate tests, for that color_to_gray.xml, at the end of "output" tag, I pressed ctrl+alt+t and I get 16 same test template....also, it seems 'repeat' tag within 'conditional' in the generated test template is currently not yet supported

thanks so much..

**Comment by davelopez** [comment]
Thanks for sharing the tool. It helped me a lot to understand better the issues :)

>The plugin version is 0.3.2 as it is the default in the vs code markets.

Yes, that is the latest version, thanks for confirming.

>If I open an existing tool wrapper, and trying to see if the tool is valid, after putting my mouse cursor after <param and press space I was expecting to see the same list of valid attributes for <param, but instead, I see a different list "id, version, hidden, display_interface" etc... I thought this list is actually the attributes for tool, and thus caused some confusion.

Ok, I think I found the problem here. You probably were talking about the `<param>` inside the `<xml>` macros. That is indeed not supported at the moment because macros are not completely defined in the Galaxy XSD schema. Also, it is a bug since it shows, as you said, the `tool` attributes when it shouldn't show any attribute autocompletion at all. I have opened an issue to track this #107 thanks!

>For the auto-generate tests, for that color_to_gray.xml, at the end of "output" tag, I pressed ctrl+alt+t and I get 16 same test template....

If you look carefully at each test you may notice that **even if they look similar**, they represent a different combination of the conditionals options (pay special attention to the param value right after each conditional):

````
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Combine"/>
            <param name="name_output_image" value="OrigGray"/>
            <conditional name="con_image_type">
                <param name="image_type" value="RGB"/>
                <param name="weight_red_channel" value="1.0"/>
                <param name="weight_green_channel" value="1.0"/>
                <param name="weight_blue_channel" value="1.0"/>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Combine"/>
            <param name="name_output_image" value="OrigGray"/>
            <conditional name="con_image_type">
                <param name="image_type" value="HSV"/>
                <param name="weight_red_channel" value="1.0"/>
                <param name="weight_green_channel" value="1.0"/>
                <param name="weight_blue_channel" value="1.0"/>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Combine"/>
            <param name="name_output_image" value="OrigGray"/>
            <conditional name="con_image_type">
                <param name="image_type" value="Channels"/>
                <repeat name="rpt_channel">
                    <param name="channel_no" value="1"/>
                    <param name="weight_of_channel" value="1.0"/>
                </repeat>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="RGB"/>
                <conditional name="con_convert_red">
                    <param name="yes_no" value="Yes"/>
                    <param name="name_output_image" value="OrigRed"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="RGB"/>
                <conditional name="con_convert_red">
                    <param name="yes_no" value="No"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="RGB"/>
                <conditional name="con_convert_green">
                    <param name="yes_no" value="Yes"/>
                    <param name="name_output_image" value="OrigGreen"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="RGB"/>
                <conditional name="con_convert_green">
                    <param name="yes_no" value="No"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="RGB"/>
                <conditional name="con_convert_blue">
                    <param name="yes_no" value="Yes"/>
                    <param name="name_output_image" value="OrigBlue"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="RGB"/>
                <conditional name="con_convert_blue">
                    <param name="yes_no" value="No"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="HSV"/>
                <conditional name="con_convert_hue">
                    <param name="yes_no" value="Yes"/>
                    <param name="name_output_image" value="OrigHue"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="HSV"/>
                <conditional name="con_convert_hue">
                    <param name="yes_no" value="No"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="HSV"/>
                <conditional name="con_convert_saturation">
                    <param name="yes_no" value="Yes"/>
                    <param name="name_output_image" value="OrigSaturation"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="HSV"/>
                <conditional name="con_convert_saturation">
                    <param name="yes_no" value="No"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="HSV"/>
                <conditional name="con_convert_value">
                    <param name="yes_no" value="Yes"/>
                    <param name="name_output_image" value="OrigValue"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="HSV"/>
                <conditional name="con_convert_value">
                    <param name="yes_no" value="No"/>
                </conditional>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
    
    <test expect_num_outputs="">
        <!--TODO: auto-generated test case. Please fill in the required values-->
        <param name="input_pipeline" value=""/>
        <param name="name_input_image" value="DNA"/>
        <conditional name="con_conversion_method">
            <param name="conversion_method" value="Split"/>
            <conditional name="con_image_type">
                <param name="image_type" value="Channels"/>
                <repeat name="rpt_channel">
                    <param name="channel_no" value="1"/>
                    <param name="image_name" value=""/>
                </repeat>
            </conditional>
        </conditional>
        <output name="output_pipeline">
            <assert_contents>
                <has_text text=""/>
                <has_line line=""/>
                <has_line_matching expression=""/>
                <has_n_columns n=""/>
                <has_size value="" delta=""/>
            </assert_contents>
        </output>
    </test>
````

>it seems 'repeat' tag within 'conditional' in the generated test template is currently not yet supported

I think they are supported, look, for example right above at the last test case inside the conditional `con_image_type`. There is a `<repeat name="rpt_channel">`. Notice that the `repeat` only shows in the tests which `con_image_type.image_type == Channels`.

Again many thanks for your feedback it is really useful! :)

**Comment by sunyi000** [comment]
sorry I overlooked the tests and repeats...

I just experience something else, not sure it is because of my misuse of the plugin or a bug...appreciate if you could have a look..

In my VS code, I have color_to_gray.xml open, then
1. I closed it -> opened a new file track_objects.xml
2. put cursor at the outputs tag in track_objects.xml and press ctrl+alt+t
VS code automatically open color_to_gray.xml and added the tests to color_to_gray.xml

then I closed every files in VS code, as well as VS itself. 
1. start VS code again.
2. File -> open file, to open track_objects.xml
3. put my cursor at the outputs tag in track_objects.xml and press ctr+alt+t
nothing happened, no tests were generated.

I repeated a couple of times the above steps, and outcome are the same.. is there something I missed here?

Thanks,

**Comment by davelopez** [comment]
>In my VS code, I have color_to_gray.xml open, then
>1. I closed it -> opened a new file track_objects.xml
>2. put cursor at the outputs tag in track_objects.xml and press ctrl+alt+t
VS code automatically open color_to_gray.xml and added the tests to color_to_gray.xml

Sorry I couldn't reproduce this behavior. Could it be some similar keyboard-shortcut that triggers the `open recent` file in VSCode? You can check your keyboard shortcuts in `File → Preferences → Keyboard Shortcuts` and remap the `galaxytools.generate.tests` command to some keyboard-shortcut that doesn't interfere with others that you may have.

>3. put my cursor at the outputs tag in track_objects.xml and press ctr+alt+t
nothing happened, no tests were generated.

This is actually a double bug :)
The first bug is in the test case generation that causes an unexpected exception and the second bug is that there is no error message when the generation fails 😅 
The good news is that it is really easy to fix and hard to find and you managed to find it :)
The bad news is that after fixing it, it generates 43 tests cases, which means there are 43 possible conditional paths in your `track_objects` tool that you may potentially want to test 😅 

Thanks again @sunyi000!

**Comment by davelopez** [comment]
I will close this, as most of the problems found should be fixed in the new [release v0.4.0](https://github.com/galaxyproject/galaxy-language-server/releases/tag/v0.4.0).

Feel free to open any new issues if you encounter any problems :)
